### PR TITLE
Add delegate support for command reporting

### PR DESCRIPTION
**Pull request by neonichu**
Commands can now report additional errors, notes and warnings, not tied to a command process as before. This is used by the `StaleFileRemoval` command to report deleted files and demotes all its errors to warnings.

See <https://github.com/apple/swift-llbuild/pull/147#discussion_r116107807>

### CODE REVIEW

**Comment by neonichu** [comment]
@swift-ci please smoke test

File: include/llbuild/BuildSystem/BuildSystem.h [hunk 1]
```diff
@@ -112,6 +112,11 @@ class BuildSystemDelegate {
                      const Token& at,
                      const Twine& message) = 0;
 
+  /// Called to provide informational output of the build system.
+  ///
+  /// \param message The informational message.
+  virtual void output(const Twine& message) = 0;
```
Comment (ddunbar) [inline include/llbuild/BuildSystem/BuildSystem.h]: I don't think we want this at the BuildSystem level, do we? Shouldn't we have it on a per-command basis, similar to `commandProcessHadOutput`?

**Comment by neonichu** [comment]
Not sure, I did this to mirror `error()`, but you're probably right.

**Comment by neonichu** [comment]
@ddunbar Do you think we should move this? If we do, we would have to make it possible for regular commands to get the execution queue delegate, that's where the command-based delegate methods are located right now.

**Comment by ddunbar** [comment]
I do think we should move it -- we want the output to be specific to a particular command. I haven't looked yet at what it would take to accomplish that.

**Comment by neonichu** [comment]
It wasn't actually necessary to provide access to the execution delegate as there was no reason to colocate the new delegate methods with the command process related ones.

PR isn't quite done, yet, I'm investigating a test issue where one file isn't reporting correctly.

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
@swift-ci please smoke test

File: lib/BuildSystem/BuildSystem.cpp [hunk 2]
```diff
@@ -2444,25 +2444,27 @@ class StaleFileRemovalCommand : public Command {
...
-        bsci.getDelegate().error("", {}, (Twine("Stale file '") + fileToDelete + "' has a relative path. This is invalid in combination with the root path attribute."));
+        bsci.getDelegate().commandHadWarning(this, "Stale file '" + fileToDelete + "' has a relative path. This is invalid in combination with the root path attribute.");
         continue;
       }
 
       // Check if the file is located under one of the allowed root paths.
       for (auto root : roots) {
         auto res = std::mismatch(root.begin(), root.end(), fileToDelete.begin());
-        if (res.first == root.end() && ((*(res.first++) == '\0') || (*(res.first++) == path_separator))) {
+        if (res.first == root.end() && ((*(res.second++) == '\0') || (*(res.second++) == path_separator))) {
```
Comment (neonichu) [inline lib/BuildSystem/BuildSystem.cpp]: This was a pretty bad bug, good that we added an additional test :)

**Comment by neonichu** [comment]
Should be good to go now, there was a bug in the test (but also in the code).

**Comment by neonichu** [comment]
Looks like the order of `filesToDelete` ends up being different on Linux for this test, I can reproduce this locally:

```
[ RUN      ] BuildSystemTaskTests.staleFileRemovalWithRoots
/home/buildnode/disk2/workspace/swift-llbuild-PR-Linux-smoke-test/llbuild/unittests/BuildSystem/BuildSystemTaskTests.cpp:499: Failure
Value of: delegate.getMessages()
  Actual: { "commandPreparing(C.1)", "commandStarted(C.1)", "commandWarning(C.1) Stale file '/bar/a.out' is located outside of the allowed root paths.", "commandWarning(C.1) Stale file '/foobar.txt' is located outside of the allowed root paths.", "commandWarning(C.1) cannot remove stale file '/foo': No such file or directory", "commandFinished(C.1)" }
Expected: std::vector<std::string>({ "commandPreparing(C.1)", "commandStarted(C.1)", "commandWarning(C.1) Stale file '/foobar.txt' is located outside of the allowed root paths.", "commandWarning(C.1) cannot remove stale file '/foo': No such file or directory", "commandWarning(C.1) Stale file '/bar/a.out' is located outside of the allowed root paths.", "commandFinished(C.1)", })
Which is: { "commandPreparing(C.1)", "commandStarted(C.1)", "commandWarning(C.1) Stale file '/foobar.txt' is located outside of the allowed root paths.", "commandWarning(C.1) cannot remove stale file '/foo': No such file or directory", "commandWarning(C.1) Stale file '/bar/a.out' is located outside of the allowed root paths.", "commandFinished(C.1)" }
[  FAILED  ] BuildSystemTaskTests.staleFileRemovalWithRoots (101 ms)
```

Will update the test.

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
The order seems to be stable per platform, so I opted for an `#if` in the test instead of making the comparison agnostic to the ordering.

**Comment by ddunbar** [comment]
@neonichu I don't think that is safe... it is probably file system dependent. Can we sort the results before printing them instead? We should always default to doing something which is deterministic where possible.

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
@ddunbar Updated the test to sort the messages before comparison. Don't think the filesystem plays into this, though, there has to be a difference in `std::set_difference` or so between the platforms.

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
Did a slight update to the messages: they should end with a newline and the successful removal message had a typo.

**Comment by neonichu** [comment]
Rebased to resolve conflicts with current `master`.

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
@swift-ci please smoke test

File: lib/BuildSystem/BuildSystem.cpp (hunk 2 above)
Comment (ddunbar) [inline lib/BuildSystem/BuildSystem.cpp]: How is this related to the other PR that changes this line?

Comment (neonichu) [inline lib/BuildSystem/BuildSystem.cpp]: It was discovered by a test which wasn't running before, because it relied on the command not failing immediately. This was the commented assertion with the `FIXME` in the `staleFileRemovalWithRoots` test.

File: lib/BuildSystem/BuildSystem.cpp [hunk 3]
```diff
@@ -2444,40 +2444,38 @@ class StaleFileRemovalCommand : public Command {
     computeFilesToDelete();
 
     bsci.getDelegate().commandStarted(this);
-    auto success = true;
```
Comment (neonichu) [inline lib/BuildSystem/BuildSystem.cpp]: Not needed anymore, because the command will not fail if files cannot be removed.

File: unittests/BuildSystem/BuildSystemTaskTests.cpp [hunk 4]
```diff
@@ -408,8 +408,8 @@ TEST(BuildSystemTaskTests, staleFileRemoval) {
     "commandPreparing(C.1)",
     "commandStarted(C.1)",
     // FIXME: Maybe it's worth creating a virtual FileSystem implementation and checking if `remove` has been called
-    "cannot remove stale file 'a.out': No such file or directory",
-    "commandFinished(C.1: 1)",
+    "commandWarning(C.1) cannot remove stale file 'a.out': No such file or directory\n",
+    "commandFinished(C.1: 0)",
```
Comment (neonichu) [inline unittests/BuildSystem/BuildSystemTaskTests.cpp]: Just as a note, all these commands are now successful, since we demoted the messages to warnings.

**Comment by neonichu** [comment]
Rebased this on top of current `master` because of conflicts.

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
@swift-ci please smoke test
